### PR TITLE
5758 for Go

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes the Go part of #5758 (reconstructing from a stale PR that introduced changes across all languages)

- [ ] Check that pulumi-azure-native compiles with the changes
- [ ] Check that pulumi-aws compiles with the changes
- [x] Understand why "listStorageAccountKeys" test is failing 
- [x] Automatically check the generated code compiles

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by t0yv0** [comment]
Ok I would not merge this until we know azure-native compiles, this is waiting on https://github.com/pulumi/pulumi-azure-native/pull/1077 I think but we can start reviewing.

File: pkg/codegen/go/gen_test.go [hunk 1]
```diff
@@ -250,3 +259,88 @@ func TestEnumUsage(t *testing.T) {
...
+		cmd.Dir = goDir
+		assert.NoError(t, cmd.Run())
+
+		t.Logf("cd %s && go build .", goDir)
+		cmd = exec.Command("go", "build", ".")
+		cmd.Dir = goDir
+		assert.NoError(t, cmd.Run())
+	})
+
+	t.Run("testGeneratedCode", func(t *testing.T) {
```
Comment (justinvp) [inline pkg/codegen/go/gen_test.go:340]: Nice

File: pkg/codegen/internal/test/testdata/output-funcs/go/codegen_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+		assert.Equal(t, "key", keys[0].KeyName)
+		assert.Equal(t, "permissions", keys[0].Permissions)
+		assert.Equal(t, "accountName=my-account-name, resourceGroupName=my-resource-group-name, expand=my-expand",
+			keys[0].Value)
+
+		return nil
+	})
+}
+
+// TODO: it seems that default values are not supported by Go codegen
```
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/output-funcs/go/codegen_test.go:152]: File a tracking issue and link to it here?

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: pkg/codegen/internal/test/testdata/output-funcs/go/codegen_test.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
```
Comment (justinvp) [inline pkg/codegen/internal/test/testdata/output-funcs/go/codegen_test.go:152]: Nit: We tend to format these like:

```go
// TODO[pulumi/pulumi#7811]: 
```

Comment (t0yv0) [reply pkg/codegen/internal/test/testdata/output-funcs/go/codegen_test.go:152]: Will do, very curious is there a tool that consumes these?

**Comment by t0yv0** [comment]
I'm going to merge this in as most essential checks are sometimes green (when not flaky).. it's been tough to get a fully green checkmark esp since Python linting is broken on master (fixing separately).

**Pull request merged**